pandas/core/reshape/pivot.py

```diff
@@ -503,20 +511,19 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
     if values is not None and aggfunc is None:
         raise ValueError("values cannot be used without an aggfunc.")
 
+    df = DataFrame(data, index=common_idx)
     if values is None:
-        df = DataFrame(data)
         df['__dummy__'] = 0
-        table = df.pivot_table('__dummy__', index=rownames, columns=colnames,
-                               aggfunc=len, margins=margins,
-                               margins_name=margins_name, dropna=dropna)
-        table = table.fillna(0).astype(np.int64)
-
+        aggfunc = len
     else:
-        data['__dummy__'] = values
-        df = DataFrame(data)
-        table = df.pivot_table('__dummy__', index=rownames, columns=colnames,
-                               aggfunc=aggfunc, margins=margins,
-                               margins_name=margins_name, dropna=dropna)
+        df['__dummy__'] = values
+
+    table = df.pivot_table('__dummy__', index=rownames, columns=colnames,
+                           aggfunc=aggfunc, margins=margins,
+                           margins_name=margins_name, dropna=dropna)
+
+    if values is None:
+        table = table.fillna(0).astype(np.int64)
 
```

Comment:
It casted back to int when there are empty cells (which are originally NaNs). I have changed the call to ``pivot_table`` so that when #17013 is fixed we can removed these two lines.